insights/parsers/ip.py

```diff
@@ -132,22 +143,34 @@ def parse_inet(line, d):
 
 def parse_rx_stats(line, d):
     split_content = line.split()
-    d["rx_bytes"] = int(split_content[0])
-    d["rx_packets"] = int(split_content[1])
-    d["rx_errors"] = int(split_content[2])
-    d["rx_dropped"] = int(split_content[3])
-    d["rx_overrun"] = int(split_content[4])
-    d["rx_mcast"] = int(split_content[5])
+    if len(split_content) == 4:  # for Virtual functions
```

Comment:
I have one query about this check: Does this `len(split_content) == 4` means:  only the `RX` line which has and only has `4` split-parts is a part of the `vf` line (Virtual functions)? For any `RX` lines which have `non-4` lines, even it is after the `vf` line, it is not the part of the `vf`?


Since, from the description of #2366, I can get 
```
    RX: bytes  packets  errors  dropped overrun mcast   
    5426       87       0       0       0       87      
    TX: bytes  packets  errors  dropped carrier collsns 
    0          0        0       0       0       0       
    vf 0 MAC 5a:17:08:5a:ae:c9, spoof checking on, link-state auto, trust off
    RX: bytes  packets  mcast   bcast   dropped       <--<< "5" parts here
    0          0        0       0       0        
    TX: bytes  packets   dropped                       <--<< "3" parts here
    0          0        0        
```
If the `RX` line of `vf` can contains more parts than `4`, I don't think the current code can handle the above output snippet in #2366 correctly.

Better make the check more general.

Reply:
@xiangce 
I got a sosreport from @yadneshk where the RX and TX lines has 5 and 3 parts respectively. The data in the test file is also from one of the sosreport where RX and TX lines has 4 and 2 parts respectively. So I guess an additional `dropped` field appears in certain scenarios. Need to handle such cases.

What do you mean by "generic" changes in "Better make the check more general."? Do you have any suggestions?

Reply:
In this case, the length of the `RX` line is variable, here we just found it's variable for "VF", but it might be variable for all cases too. It's not a good idea to parse it in a `fixed` pattern.

Here is a `generic` method I can imagine:

```diff
diff --git a/insights/parsers/ip.py b/insights/parsers/ip.py
index 7c3ac82a..fff87cf4 100644
--- a/insights/parsers/ip.py
+++ b/insights/parsers/ip.py
@@ -55,16 +55,16 @@ def parse_ip_addr(content):
     r = {}
     current = {}
     if_details = {}
-    rx_next_line = False
-    tx_next_line = False
+    rx_next_line = []
+    tx_next_line = []
     content = [l.strip() for l in content if "Message truncated" not in l]
     for line in filter(None, content):
         if rx_next_line and current:
-            parse_rx_stats(line, current)
-            rx_next_line = False
+            parse_rx_stats(line, current, rx_next_line)
+            rx_next_line = []
         if tx_next_line and current:
-            parse_tx_stats(line, current)
-            tx_next_line = False
+            parse_tx_stats(line, current, tx_next_line)
+            tx_next_line = []
         elif line[0].isdigit() and "state" in line:
             current = parse_interface(line)
             r[current["name"]] = current
@@ -81,9 +81,9 @@ def parse_ip_addr(content):
         elif line.startswith("inet"):
             parse_inet(line, current)
         elif line.startswith("RX"):
-            rx_next_line = True
+            rx_next_line = line.split()[1:]
         elif line.startswith("TX"):
-            tx_next_line = True
+            tx_next_line = line.split()[1:]
     for k, v in r.items():
         if_details[k] = NetworkInterface(v)
     return if_details
@@ -141,36 +141,16 @@ def parse_inet(line, d):
     })


-def parse_rx_stats(line, d):
+def parse_rx_stats(line, d, rx_keys):
     split_content = line.split()
-    if len(split_content) == 4:  # for Virtual functions
-        d = d['vf'][-1]
-        d["rx_bytes"] = int(split_content[0])
-        d["rx_packets"] = int(split_content[1])
-        d["rx_mcast"] = int(split_content[2])
-        d["rx_bcast"] = int(split_content[3])
-    else:
-        d["rx_bytes"] = int(split_content[0])
-        d["rx_packets"] = int(split_content[1])
-        d["rx_errors"] = int(split_content[2])
-        d["rx_dropped"] = int(split_content[3])
-        d["rx_overrun"] = int(split_content[4])
-        d["rx_mcast"] = int(split_content[5])
+    rx_keys = ['rx_{0}'.format(k) for k in rx_keys]
+    d.update(dict(zip(rx_keys), split_content))


-def parse_tx_stats(line, d):
+def parse_tx_stats(line, d, tx_keys):
     split_content = line.split()
-    if len(split_content) == 2:  # for Virtual functions
-        d = d['vf'][-1]
-        d["tx_bytes"] = int(split_content[0])
-        d["tx_packets"] = int(split_content[1])
-    else:
-        d["tx_bytes"] = int(split_content[0])
-        d["tx_packets"] = int(split_content[1])
-        d["tx_errors"] = int(split_content[2])
-        d["tx_dropped"] = int(split_content[3])
-        d["tx_carrier"] = int(split_content[4])
-        d["tx_collsns"] = int(split_content[5])
+    tx_keys = ['tx_{0}'.format(k) for k in tx_keys]
+    d.update(dict(zip(tx_keys), split_content))

```

Reply:
How is it going? Any update?

Reply:
any update?

Reply:
Hi. I'm working on it.